WIP: fixing suggested commands for js versions

garikbesson opened the pull request.

frol commented on src/js_command_match/add_credentials.rs:
```diff
@@ -0,0 +1,97 @@
+
+impl AddCredentialsArgs {
+    pub fn to_cli_args(&self, network_config: String) -> Vec<String> {
+        let network_id = self.network_id.clone().unwrap_or(network_config.to_owned());
```
This `.to_owned()` is unnecessary as it is already an owned `String` (not a reference):

```suggestion
        let network_id = self.network_id.clone().unwrap_or(network_config);
```

frol commented on src/js_command_match/add_credentials.rs:
```diff
@@ -0,0 +1,97 @@
+            seed_phrase: Some(seed_phrase.clone()),
+            secret_key: None,
+            network_id: None,
+            _unknown_args: [].to_vec(),
```
Here is a more ergonomic way to write it:

```suggestion
            _unknown_args: vec![],
```

frol commented on src/js_command_match/add_key.rs:
```diff
@@ -3,49 +3,116 @@
     account_id: String,
     access_key: String,
-    #[clap(long, aliases = ["contract_id", "contractId"], default_value = None)]
+    #[clap(long, aliases = ["contractId"], default_value = None)]
```
I vaguely remember that JS CLI actually supported three ways of passing flags: camelCase (`--contractId`), kebab-case (`--contract-id`), and snake_case (`--contract_id`). Are you sure that we don't need to support snake_case?

frol commented:
Please, resolve the conflicts and rewrite integration tests into Rust:

If you are new to integration tests in Rust, Rust book has a great chapter: https://doc.rust-lang.org/book/ch11-03-test-organization.html#integration-tests (tip: don't miss the section called "Integration Tests for Binary Crates")

P.S. If you feel it is too much, let's call for help to rewrite the tests into Rust, just let me know

garikbesson commented:
@frol Hi, could you review the PR again?

That's key things that have changed since the last review:

1. I removed end-to-end tests written by me in JavaScript
2. I implemented unit tests that parse args and check how the interpreted commands look like
3. I skipped some following methods that are deprecated on JS CLI: `clean`, `dev-deploy`, `evm-call`, `evm-view`, `evm-dev-init`, `js`, `repl`, `set-api-key`
4. I skipped validator-related methods: `proposals`, `stake`, `validators`
5. All interpreted commands now leverage `sign-with-keychain` option (instead of leveraging the legacy keychain)

CC: @gagdiez

gagdiez commented:
All relevant commands are now implemented:

- account: create, delete, login, check state
- contract: call, deploy, check storage, view
- keys: add, delete, list keys
- transactions: send near, check status

The following were not ported since they were deprecated for months:
- evm commands, dev-deploy, repl, clean

Validators now simply shows a message asking to use the extension, without explaining how to use the extension

gagdiez reviewed (approved):
I have tested all codes and its variants, and the proposed commands work accordingly

frol commented on src/js_command_match/transactions/status.rs:
```diff
@@ -0,0 +1,70 @@
+        let transaction_hash = "4HxfV69Brk7fJd3NC63ti2H3QCgwiUiMAPvwNmGWbVXo";
+        let network_id = "mainnet";
+
+        for i in 0..NETWORK_ID_ALIASES.len() {
+            let network_id_parameter_alias = &format!("--{}", &NETWORK_ID_ALIASES[i]);
+            let state_args = TxStatusArgs::parse_from(&[
+                "near",
+                transaction_hash,
+                network_id_parameter_alias,
+                network_id,
+            ]);
```
nit: You can simplify it to (and apply in other places):

```suggestion
        for network_id_alias in NETWORK_ID_ALIASES {
            let state_args = TxStatusArgs::parse_from(&[
                "near",
                transaction_hash,
                &format!("--{}", &network_id_alias),
                network_id,
            ]);
```

frol commented on src/js_command_match/transactions/status.rs:
```diff
@@ -0,0 +1,70 @@
+#[derive(Debug, Clone, clap::Parser)]
+pub struct TxStatusArgs {
+    hash: String,
+    #[clap(allow_hyphen_values = true, num_args = 0..)]
+    _unknown_args: Vec<String>,
+    #[clap(long, aliases = NETWORK_ID_ALIASES, default_value=None)]
+    network_id: Option<String>,
```
Is there a specific meaning for putting `_unknown_args` in the middle instead of the end of the fields? It not, let's move it down:

```suggestion
    #[clap(long, aliases = NETWORK_ID_ALIASES, default_value=None)]
    network_id: Option<String>,
    #[clap(allow_hyphen_values = true, num_args = 0..)]
    _unknown_args: Vec<String>,
```

frol commented on src/js_command_match/keys/list.rs:
```diff
@@ -0,0 +1,71 @@
+use crate::js_command_match::constants::NETWORK_ID_ALIASES;
+
+#[derive(Debug, Clone, clap::Parser)]
+#[clap(alias("list-keys"))]
```
Since the command in the top-level enum included as `ListKeys`, it is already expecting `list-keys` command, so this alias only causes a parsing error:

```
env RUST_BACKTRACE=1 cargo run -- validators
```

<img width="1082" alt="image" src="https://github.com/near/near-cli-rs/assets/304265/a5d86e62-cfd9-430d-b09c-886c052c22ad">

Consider removing the alias or if there is a different alias, use it here.

```suggestion
```

frol commented on src/js_command_match/keys/add.rs:
```diff
@@ -0,0 +1,169 @@
+            let result = AddKeyArgs::to_cli_args(&add_key_args, "testnet".to_string());
+            assert_eq!(
+                result.join(" "),
+                format!("account add-key {} grant-function-call-access --allowance unlimited --contract-account-id {} --function-names {} use-manually-provided-public-key {} network-config testnet sign-with-keychain send", account_id, contract_id, method_names, access_key)
```
nit: It might be easier to read this way:

```suggestion
                format!("account add-key {account_id} grant-function-call-access --allowance unlimited --contract-account-id {contract_id} --function-names {method_names} use-manually-provided-public-key {access_key} network-config testnet sign-with-keychain send")
```

frol commented on src/js_command_match/contract/call.rs:
```diff
@@ -0,0 +1,179 @@
+        command.push("network-config".to_string());
+        command.push(network_id);
+
+        if self.sign_with_ledger {
+            command.push("sign-with-ledger".to_string());
+            command.push("--seed-phrase-hd-path".to_string());
+            command.push(self.ledger_path.to_owned().unwrap_or_default());
+        }
+
+        if self.private_key.is_some() {
+            command.push("sign-with-plaintext-private-key".to_string());
+            command.push(self.private_key.clone().unwrap());
+        }
+
+        if !self.sign_with_ledger && self.private_key.is_none() {
+            command.push("sign-with-keychain".to_string());
+        }
```
nit: Consider using `else` and `if let` instead of `if` + `.unwrap()`:

```suggestion
        if self.sign_with_ledger {
            command.push("sign-with-ledger".to_string());
            command.push("--seed-phrase-hd-path".to_string());
            command.push(self.ledger_path.to_owned().unwrap_or_default());
        } else if let Some(private_key) = self.private_key {
            command.push("sign-with-plaintext-private-key".to_string());
            command.push(private_key.clone());
        } else {
            command.push("sign-with-keychain".to_string());
        }
```

frol commented on src/js_command_match/contract/call.rs:
```diff
@@ -0,0 +1,179 @@
+        assert_eq!(
+                result.join(" "),
+                format!(
+                    "contract call-function as-transaction {} flip_coin json-args {} prepaid-gas 30000000000000 gas attached-deposit 0 NEAR sign-as {} network-config testnet sign-with-keychain send",
```
I am surprised it passes on CI and on your side. It fails for me:

<img width="1508" alt="image" src="https://github.com/near/near-cli-rs/assets/304265/9f76d6f3-8ce6-4105-a49c-d93580aeaafd">

Could you take a look? I expect it to render as `30 Tgas`, not `30000000000000 gas`, so my local test fails as I expect it.

gagdiez replied:
indeed, it was failing, thanks for catching it!

gagdiez commented on src/js_command_match/account/create.rs:
```diff
@@ -0,0 +1,265 @@
+        if self.use_ledger_pk {
+            command.push("use-ledger".to_owned());
+
+            // add after issue with ledger key is resolved
+            // command.push(format!("--seed-phrase-hd-path {}", self.pk_ledger_path.clone().unwrap()));
```
https://github.com/near/near-cli-rs/issues/361

frol replied:
@gagdiez That is resolved now. Please, check and update this code accordingly

garikbesson replied:
That was updated

frol commented on src/js_command_match/constants.rs:
```diff
@@ -0,0 +1,37 @@
+    "useAccount",
+    "use-account",
+    "accountId",
+    "account_id",
```
I believe it should have hyphen here, instead of underscore:

```suggestion
    "account-id",
```

Otherwise, I get an error for a valid JS CLI command:

<img width="963" alt="image" src="https://github.com/near/near-cli-rs/assets/304265/33977780-1a51-46e4-80e5-a9159dc0180a">

<img width="961" alt="image" src="https://github.com/near/near-cli-rs/assets/304265/2a801322-3728-4272-8e36-fc609d24eeb7">

garikbesson replied:
underscrore's versions were removed from each of aliases

frol commented on src/js_command_match/constants.rs:
```diff
@@ -0,0 +1,37 @@
+
+// SETTINGS
+pub const NETWORK_ID_ALIASES: [&str; 2] = ["networkId", "network-id"];
+pub const BLOCK_ID_ALIASES: [&str; 2] = ["blockId", "block_id"];
+
+// Deploy
+pub const WASM_FILE_ALIASES: [&str; 3] = ["wasm_file", "wasmFile", "wasm-file"];
+pub const INIT_FUNCTION_ALIASES: [&str; 3] = ["init_function", "initFunction", "init-function"];
+pub const INIT_ARGS_ALIASES: [&str; 3] = ["init_args", "initArgs", "init-args"];
+pub const INIT_GAS_ALIASES: [&str; 3] = ["init_gas", "initGas", "init-gas"];
+pub const INIT_DEPOSIT_ALIASES: [&str; 3] = ["init_deposit", "initDeposit", "init-deposit"];
```
Should `_` be `-` here?

garikbesson replied:
underscrore's versions were removed from each of aliases

garikbesson commented:
@race-of-sloths please, include my PR in the Race

race-of-sloths commented:
@garikbesson Thank you for your contribution! Your pull request is now a part of the [Race of Sloths](https://race-of-sloths.com)!
New Sloth joined the Race! Welcome!

[<picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://badge.race-of-sloths.com/garikbesson?type=bot&theme=dark">
    <source media="(prefers-color-scheme: light)" srcset="https://badge.race-of-sloths.com/garikbesson?type=bot&theme=light">
    <img alt="Shows profile picture for the author of the PR" src="https://badge.race-of-sloths.com/garikbesson?type=bot&theme=white">
</picture>](https://race-of-sloths.com/profile/garikbesson)


<details><summary>Current status: <i>waiting for scoring</i></summary>

We're waiting for maintainer to score this pull request with `@race-of-sloths score [0,1,2,3,5,8,13]` command
</details><details><summary>What is the Race of Sloths</summary>
<p>
Race of Sloths is a friendly competition where you can participate in challenges and compete with other open-source contributors within your normal workflow

For contributors:

- Tag @race-of-sloths inside your pull requests
- Wait for the maintainer to review and score your pull request
- Check out your position in the [Leaderboard](https://race-of-sloths.com/leaderboard)
- Keep weekly and monthly streaks to reach higher positions
- Boast your contributions with a dynamic picture of your [Profile](https://race-of-sloths.com/profile/garikbesson)

For maintainers:
- Score pull requests that participate in the Race of Sloths
- Engage contributors with fair scoring and fast responses so they keep their streaks
- Promote the Race to the point where the Race starts promoting you
- Grow the community of your contributors

Feel free to check [our website](https://race-of-sloths.com) for additional details!
</p>
</details> 

<details>
<summary>Bot commands</summary>

- For contributors
  - **Include a PR:** `@race-of-sloths include` to enter the Race with your PR
- For maintainers:
  - **Assign points:** `@race-of-sloths score [1/2/3/5/8/13]` to award points based on your assessment.
  - **Reject this PR:** `@race-of-sloths exclude` to send this PR back to the drawing board.
  - **Exclude repo:** `@race-of-sloths pause` to stop bot activity in this repo until `@race-of-sloths unpause` command is called

</details>

frol reviewed (approved):
@garikbesson @gagdiez Enormous effort! Thank you for pulling it out!

@race-of-sloths score 13

race-of-sloths commented:
### 🌟 Score recorded!

@frol, thank you for scoring this pull request in the Race of Sloths!

race-of-sloths commented:
### ✅ PR is finalized! 
Your contribution is much appreciated with a final score of 13!
You have received 150 (130 base + 10 weekly bonus + 10 monthly bonus) Sloth points for this contribution


Congratulations @garikbesson! Your PR was highly scored and you completed another monthly streak! To keep your monthly streak make another pull request next month and get 8+ score for it
